### PR TITLE
[native clouds] WildFly deploy operation support for Azure

### PR DESCRIPTION
**Pull request by istraka**
@mchoma

### CODE REVIEW

File: clouds/clouds-azure/src/main/java/azure/core/AzureUtils.java [hunk 1]
```diff
@@ -50,12 +49,68 @@ static Optional<VirtualMachine> findAzureVM(AzureResourceManager arm, String nam
...
+        while (System.currentTimeMillis() - start < timeout) {
+            try {
+                Response response = client.newCall(new Request.Builder()
+                                .url("http://" + app.defaultHostname())
+                                .method("GET", null)
+                                .build())
+                        .execute();
+                int code = response.code();
+                String body = response.body() != null ? response.body().string() : null;
+                if (code < 500 && !isWebAppDnsProblem(code, body) && !isEapWebAppWelcomePage(code, body)) {
```
Comment (mnovak1) [inline clouds/clouds-azure/src/main/java/azure/core/AzureUtils.java:75]: Checks in `isWebAppDnsProblem(code, body) && !isEapWebAppWelcomePage` seems to be quite squishy and depend on HTML body/content. Does Azure provide some API for this?

File: clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+ *
+ * Heavily uses {@link AzureIdentifiableSunstoneResource} to determine the destination of deploy operation
+ *
+ * To retrieve Azure cloud resources, the class relies on {@link AzureIdentifiableSunstoneResource#get(Annotation, AzureSunstoneStore, Class)}.
+ *
+ * Undeploy operations are registered in the extension store so that they are closed once the store is closed
+ */
+public class AzureArchiveDeployer implements SunstoneArchiveDeployer {
+
+    static void deployToWebApp(Identification resourceIdentification, InputStream is, AzureSunstoneStore store) throws Exception {
```
[the remark this answers is not in the document]
Comment (istraka) [reply clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java:38]: No, there is only 80 port, nothing else.

File: clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        Path tempFile = Files.createTempFile("sunstone-war-deployment-", ".war");
+        Files.copy(is, tempFile, StandardCopyOption.REPLACE_EXISTING);
+        WebApp azureWebApp = resourceIdentification.get(store, WebApp.class);
+
+        Exception ex = null;
+        for (int i = 0; i < 5; i++) {
```
[the remark this answers is not in the document]
Comment (istraka) [reply clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java:44]: I had experience where it just failed. Another try was successful.

File: clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            }
+        }
+        if (ex != null) {
+            throw ex;
+        }
+        store.addClosable(() -> undeployFromWebApp(azureWebApp));
+
+        azureWebApp.restart();
+        // restart is started after some unknown time (async) :/
+        Thread.sleep(TimeoutUtils.adjust(30000));
```
[the remark this answers is not in the document]
Comment (istraka) [reply clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java:64]: It blocks indeed, but the restart is not started right away and the service is available for some time after the `azureWebApp.restart()` is done.

File: clouds/clouds-azure/src/main/java/azure/core/AzureUtils.java (hunk 1 above)
Comment (istraka) [reply clouds/clouds-azure/src/main/java/azure/core/AzureUtils.java:75]: I am not aware of any. I will ping Denis from MS team if he knows something.

File: clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mnovak1) [reply clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java:64]: I see. I'm thinking whether Azure might provide restart counter or some other way checking the state which might be watched to see that restart was triggered. Could you check it, please?

File: clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (mnovak1) [reply clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java:44]: Ok, not nice but we will have to live with it.

File: clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (mnovak1) [reply clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java:38]: Ok, I vaguely recall it's the recommended way of deploying WAR, right?

File: clouds/clouds-azure/pom.xml [hunk 5]
```diff
@@ -29,6 +29,13 @@
             <version>${project.version}</version>
         </dependency>
 
+        <!-- https://mvnrepository.com/artifact/commons-net/commons-net -->
```
Comment (mchoma) [inline clouds/clouds-azure/pom.xml:32]: Is this comment important?

File: clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (mchoma) [reply clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java:64]: We need something like Waiter?

That reminds me situation in OpenShift, where also it is necessary to wait for things to happen.

File: clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/vm/suitetests/AzureVmUndeployedSecondTest.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+public class AzureVmUndeployedSecondTest {
+    @AzureVirtualMachine(name = instanceName, group = vmDeployGroup)
+    Hostname hostname;
+
+    @Test
+    public void test() throws IOException {
+        OkHttpClient client = new OkHttpClient();
+
+        Request request = new Request.Builder()
+                .url("http://" + hostname.get() + ":8080/testapp")
```
Comment (mchoma) [inline clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/vm/suitetests/AzureVmUndeployedSecondTest.java:34]: So we rely here on fact app `testapp` was deployed in previous test???

File: clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/webapp/WebAppDeploySuiteTests.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+
+import azure.armTemplates.archiveDeploy.webapp.suitetests.AzureWebAppDeployFirstTest;
+import azure.armTemplates.archiveDeploy.webapp.suitetests.AzureWebAppUndeployedSecondTest;
+import org.junit.platform.suite.api.SelectClasses;
+import org.junit.platform.suite.api.Suite;
+
+
+@Suite
+@SelectClasses({AzureWebAppDeployFirstTest.class, AzureWebAppUndeployedSecondTest.class})
```
Comment (mchoma) [inline clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/webapp/WebAppDeploySuiteTests.java:11]: So order in suite is important?

File: clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/webapp/suitetests/AzureWebAppDeployFirstTest.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+    @AzureWebApplication(name = instanceName, group = webAppDeployGroup)
+    Hostname hostname;
+
+    @Test
+    public void test() throws IOException {
+        OkHttpClient client = new OkHttpClient();
+
+        Request request = new Request.Builder()
+                .url("http://" + hostname.get())
```
Comment (mchoma) [inline clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/webapp/suitetests/AzureWebAppDeployFirstTest.java:41]: So when deployment name is empty it is deployed as ROOT?

What happens when I deploy multiple deployments with empty deployment name?

File: clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (istraka) [reply clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java:38]: Yes.

File: clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (istraka) [reply clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java:64]: Yes, I have been thinking whether we could benefit from Waiter API. For now, it doesn't seem it is needed anywhere else in the Sunstone. However we may discover it is needed in test development. Then We can introduce it.

File: clouds/pom.xml [hunk 9]
```diff
@@ -44,6 +44,7 @@
         <version.software.amazon.awssdk.bom>2.17.174</version.software.amazon.awssdk.bom>
 
         <version.com.google.code.gson>2.9.0</version.com.google.code.gson>
+        <version.commons-net>3.6</version.commons-net>
```
[the remark this answers is not in the document]
Comment (istraka) [reply clouds/pom.xml:47]: I will update it.

File: clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/vm/suitetests/AzureVmUndeployedSecondTest.java (hunk 6 above)
Comment (istraka) [reply clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/vm/suitetests/AzureVmUndeployedSecondTest.java:34]: We rely on the fact the Suite runs test classes sequentially. The test is supposed to be run after `AzureWebAppDeployFirstTest` and verify the artifact was undeployed correctly.

File: clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/webapp/suitetests/AzureWebAppDeployFirstTest.java (hunk 8 above)
Comment (istraka) [reply clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/webapp/suitetests/AzureWebAppDeployFirstTest.java:41]: > So when deployment name is empty it is deployed as ROOT?
No. For Az Web App it is always deployed as a ROOT.war. This is how Azure works.

File: clouds/clouds-azure/src/main/java/azure/core/AzureUtils.java (hunk 1 above)
Comment (istraka) [reply clouds/clouds-azure/src/main/java/azure/core/AzureUtils.java:75]: Denis is not very responsive. Once we will get more info, I will happily integrated suggested solution.

File: clouds/clouds-core/src/main/java/sunstone/core/SunstoneExtension.java [hunk 10]
```diff
@@ -130,6 +153,67 @@ static void handleInjection(ExtensionContext ctx, Object instance, Field field)
...
+                if (method.getParameterCount() != 0) {
+                    throw new IllegalArgumentSunstoneException("Deployment method must have 0 parameters");
+                }
+                Deployment annotation = method.getAnnotation(Deployment.class);
+                String deploymentName = ObjectProperties.replaceSystemProperties(annotation.name());
+
+                method.setAccessible(true);
+                Object invoke = method.invoke(null);
+                if (invoke == null) {
+                    throw new RuntimeException("return object is null");
```
[the remark this answers is not in the document]
Comment (istraka) [reply clouds/clouds-core/src/main/java/sunstone/core/SunstoneExtension.java]: Fixed.

File: clouds/clouds-azure/pom.xml (hunk 5 above)
Comment (istraka) [reply clouds/clouds-azure/pom.xml]: Removed.

File: clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (istraka) [reply clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java:44]: Resolving the conversation.

File: clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+            ftpClient.login(profile.ftpUsername(), profile.ftpPassword());
+            ftpClient.setFileType(FTP.BINARY_FILE_TYPE);
+
+            FtpUtils.cleanDirectory(ftpClient, "/site/wwwroot/");
+
+            ftpClient.disconnect();
+
+            webApp.restart();
+            // restart is started after some unknown time (async) :/
+            Thread.sleep(TimeoutUtils.adjust(30000));
```
[the remark this answers is not in the document]
Comment (istraka) [reply clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java:85]: See the discussion in other thread. Resolving the conversation.

File: clouds/clouds-azure/src/main/java/azure/core/AzureUtils.java [hunk 12]
```diff
@@ -50,12 +49,68 @@ static Optional<VirtualMachine> findAzureVM(AzureResourceManager arm, String nam
...
+                Response response = client.newCall(new Request.Builder()
+                                .url("http://" + app.defaultHostname())
+                                .method("GET", null)
+                                .build())
+                        .execute();
+                int code = response.code();
+                String body = response.body() != null ? response.body().string() : null;
+                // several strings from EAP Web APP welcome page
+                // expecting to be any deployment undeployed and waiting for the welcome page
+                if (isEapWebAppWelcomePage(code, body)) {
```
[the remark this answers is not in the document]
Comment (istraka) [reply clouds/clouds-azure/src/main/java/azure/core/AzureUtils.java:103]: See the conversation in the other thread. Resolving the conversation.

File: clouds/pom.xml (hunk 9 above)
[the remark this answers is not in the document]
Comment (istraka) [reply clouds/pom.xml]: Fixed.

File: clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/vm/suitetests/AzureVmUndeployedSecondTest.java (hunk 6 above)
Comment (istraka) [reply clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/vm/suitetests/AzureVmUndeployedSecondTest.java:37]: I have added a javadoc.

File: clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/webapp/suitetests/AzureWebAppDeployFirstTest.java (hunk 8 above)
Comment (istraka) [reply clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/webapp/suitetests/AzureWebAppDeployFirstTest.java:41]: I have added some checks.

**Comment by istraka** [comment]
I have fixed the issues. Please, review.

File: clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/webapp/WebAppDeploySuiteTests.java (hunk 7 above)
Comment (mchoma) [reply clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/webapp/WebAppDeploySuiteTests.java:15]: Ok

File: clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/vm/suitetests/AzureVmUndeployedSecondTest.java (hunk 6 above)
Comment (mchoma) [reply clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/vm/suitetests/AzureVmUndeployedSecondTest.java:37]: ok

File: clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/webapp/suitetests/AzureWebAppDeployFirstTest.java (hunk 8 above)
Comment (mchoma) [reply clouds/clouds-azure/src/test/java/azure/armTemplates/archiveDeploy/webapp/suitetests/AzureWebAppDeployFirstTest.java:41]: ok

File: clouds/clouds-azure/src/test/resources/azure/armTemplates/eapWebApp.json [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "appName": {
+      "type": "string",
+      "defaultValue": "eapAppService"
+    },
+    "linuxFxVersion": {
+      "type": "string",
+      "defaultValue": "JBOSSEAP|7-java11"
```
Comment (mchoma) [inline clouds/clouds-azure/src/test/resources/azure/armTemplates/eapWebApp.json:11]: For my knowledge this is actually identifying EAP image?

**Review by mnovak1 — APPROVED** [review]

**Comment by mnovak1** [comment]
LGTM, except the `Thread.sleep(..)` issue.

File: clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+            try {
+                azureWebApp.warDeploy(tempFile.toFile());
+                ex = null;
+                break;
+            } catch (Exception e) {
+                AzureLogger.DEFAULT.debug("WebApp::warDeploy try no. {} failed", i);
+                e.printStackTrace();
+                ex = e;
+                // try another time
+                Thread.sleep(10 * 1000);
```
Comment (mchoma) [inline clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java:55]: This is OK?

File: clouds/clouds-azure/src/test/resources/azure/armTemplates/eapWebApp.json (hunk 13 above)
Comment (istraka) [reply clouds/clouds-azure/src/test/resources/azure/armTemplates/eapWebApp.json:11]: Yes.

File: clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (istraka) [reply clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java]: TimeoutUtils factors the value with a configurable value.

File: clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java (hunk 14 above)
Comment (istraka) [reply clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java]: `TimeoutUtils.adjust()` should have been used. But I have got rid it using async operation and waiting for the result.

Comment (mchoma) [reply clouds/clouds-azure/src/main/java/azure/core/AzureArchiveDeployer.java]: Ok. Now I am happy :)

**Comment by mchoma** [comment]
All my objections were addressed. Thread.sleep as well. I am approving this MR.

**Review by mchoma — APPROVED** [review]

**Pull request merged**